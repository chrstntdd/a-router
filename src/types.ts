export type HistoryLocation = Window['location'] & { state?: any }

export type NavigateFn = (to: string, options?: NavigateOptions<{}>) => void

export type EventHandler = (event: Event) => void

export interface NavigateOptions<T> {
  state?: T
  replace?: boolean
}

export type HistoryListener = (p: { location: HistoryLocation; action: 'PUSH' | 'POP' }) => void
export type HistoryUnsubscribe = () => void

export interface History {
  readonly location: HistoryLocation
  readonly transitioning: boolean
  listen: (listener: HistoryListener) => HistoryUnsubscribe
  navigate: NavigateFn
}

export interface HistorySource {
  readonly location: HistoryLocation
  addEventListener(name: string, listener: EventHandler): void
  removeEventListener(name: string, listener: EventHandler): void
  history: {
    readonly state: any
    pushState(state: any, title: string, uri: string): void
    replaceState(state: any, title: string, uri: string): void
    readonly entries: any
    readonly index: number
  }
}

export type RouteComponentProps<ComponentProps = {}> = Partial<ComponentProps> & {
  path?: string
  default?: boolean
  location?: HistoryLocation
  navigate?: NavigateFn
  uri?: string
}
