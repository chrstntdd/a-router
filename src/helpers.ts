import * as React from 'react'

import { ParamsObj, HistoryLocation, NavigateFn } from './types'

const isProduction = process.env.NODE_ENV == 'production'
const prefix = 'ðŸ”¥'

/**
 * @description
 *  - Throw an error if the condition fails
 *  - Strip out error messages for production
 */
const invariant = (condition: any, message?: string): void => {
  if (condition) return

  if (isProduction) {
    throw new Error(prefix)
  }

  throw new Error(`${prefix}: ${message || ''}`)
}

const startsWith = (input: string, search: string): boolean =>
  input.substr(0, search.length) == search

const shouldNavigate = (event: React.MouseEvent<HTMLElement>) =>
  !event.defaultPrevented &&
  event.button == 0 &&
  !(event.metaKey || event.altKey || event.ctrlKey || event.shiftKey)

const stripSlashes = (str: string): string => str.replace(/(^\/+|\/+$)/g, '')

type RouteElement = {
  uri: string
  location: HistoryLocation
  navigate: NavigateFn
  children?: React.ReactElement<any>[]
}

export interface Route {
  default?: boolean
  path?: string
  value?: React.ReactElement<RouteElement>
}

interface ReturnRoute {
  params: ParamsObj
  route: Route
  uri: string
}

/**
 * @description Ranks and picks the best route to match. Each segment gets the highest
 * amount of points, then the type of segment gets an additional amount of
 * points where
 *
 * `static > dynamic > wildcard > root`
 */
const pick = (routes: Route[], uri: string): ReturnRoute | null => {
  let match
  let _default

  const [uriPathname] = uri.split('?')
  const uriSegments = segmentize(uriPathname)
  const isRootUri = uriSegments[0] == ''
  const ranked = rankRoutes(routes)

  for (let i = 0, l = ranked.length; i < l; i++) {
    const route = ranked[i].route
    let missed = false

    if (route.default) {
      _default = {
        route,
        uri,
        params: {}
      }
      continue
    }

    const routeSegments = segmentize(route.path)
    const params: ParamsObj = {}
    const max = Math.max(uriSegments.length, routeSegments.length)
    let index = 0

    for (; index < max; index++) {
      const routeSegment = routeSegments[index]
      const uriSegment = uriSegments[index]

      if (isWildcard(routeSegment)) {
        // Hit a wildcard, just grab the rest, and return a match
        // uri:   /files/documents/work
        // route: /files/*
        params['*'] = uriSegments
          .slice(index)
          .map(decodeURIComponent)
          .join('/')
        break
      }

      if (uriSegment == undefined) {
        // URI is shorter than the route, no match
        // uri:   /users
        // route: /users/:userId
        missed = true
        break
      }

      const dynamicMatch = paramRe.exec(routeSegment)

      if (dynamicMatch && !isRootUri) {
        const matchIsNotReserved = reservedNames.indexOf(dynamicMatch[1]) == -1
        invariant(
          matchIsNotReserved,
          `<Router> dynamic segment "${
            dynamicMatch[1]
          }" is a reserved name. Please use a different name in path "${route.path}".`
        )

        const value = decodeURIComponent(uriSegment)
        params[dynamicMatch[1]] = value
      } else if (routeSegment != uriSegment) {
        // Current segments don't match, not dynamic, not a wildcard, so no match
        // uri:   /users/123/settings
        // route: /users/:id/profile
        missed = true
        break
      }
    }

    if (!missed) {
      match = {
        route,
        params,
        uri: '/' + uriSegments.slice(0, index).join('/')
      }
      break
    }
  }

  return match || _default || null
}

/**
 *
 * @description Matches just one path to a uri
 */
const match = (path: string, uri: string) => pick([{ path }], uri)

/***
 * @description Resolves URIs as though every path is a directory, no files.
 * Relative URIs in the browser can feel awkward because not only can you be
 * "in a directory" you can be "at a file", too. For example
 *
 *     browserSpecResolve('foo', '/bar/') => /bar/foo
 *     browserSpecResolve('foo', '/bar') => /foo
 *
 * But on the command line of a file system, it's not as complicated, you can't
 * `cd` from a file, only directories.  This way, links have to know less about
 * their current path. To go deeper you can do this:
 *
 *     <Link to="deeper"/>
 *     // instead of
 *     <Link to=`{${props.uri}/deeper}`/>
 *
 * Just like `cd`, if you want to go deeper from the command line, you do this:
 *
 *     cd deeper
 *     # not
 *     cd $(pwd)/deeper
 *
 * By treating every path as a directory, linking to relative paths should
 * require less contextual information and (fingers crossed) be more intuitive.
 */
const resolve = (to: string, base: string) => {
  // /foo/bar, /baz/qux => /foo/bar
  if (startsWith(to, '/')) return to

  const [toPathname, toQuery] = to.split('?')
  const [basePathname] = base.split('?')

  const toSegments = segmentize(toPathname)
  const baseSegments = segmentize(basePathname)

  // ?a=b, /users?b=c => /users?a=b
  if (toSegments[0] == '') return addQuery(basePathname, toQuery)

  // profile, /users/789 => /users/789/profile
  if (!startsWith(toSegments[0], '.')) {
    const pathname = baseSegments.concat(toSegments).join('/')
    return addQuery((basePathname == '/' ? '' : '/') + pathname, toQuery)
  }

  // ./         /users/123  =>  /users/123
  // ../        /users/123  =>  /users
  // ../..      /users/123  =>  /
  // ../../one  /a/b/c/d    =>  /a/b/one
  // .././one   /a/b/c/d    =>  /a/b/c/one
  const allSegments = baseSegments.concat(toSegments)
  const segments = []
  for (let i = 0, l = allSegments.length; i < l; i++) {
    const segment = allSegments[i]
    if (segment == '..') segments.pop()
    else if (segment != '.') segments.push(segment)
  }

  return addQuery('/' + segments.join('/'), toQuery)
}

const insertParams = (path: string, params: any) => {
  const segments = segmentize(path)

  return (
    '/' +
    segments
      .map(segment => {
        const match = paramRe.exec(segment)

        return match ? params[match[1]] : segment
      })
      .join('/')
  )
}

const validateRedirect = (from: string, to: string) => {
  const filterFn = (segment: string) => isDynamic(segment)

  const fromString = segmentize(from)
    .filter(filterFn)
    .sort()
    .join('/')
  const toString = segmentize(to)
    .filter(filterFn)
    .sort()
    .join('/')

  return fromString == toString
}

////////////////////////////////////////////////////////////////////////////////
// Junk
const paramRe = /^:(.+)/

const SEGMENT_POINTS = 4
const STATIC_POINTS = 3
const DYNAMIC_POINTS = 2
const WILDCARD_PENALTY = 1
const ROOT_POINTS = 1

const isRootSegment = (segment: string) => segment == ''

const isDynamic = (segment: string) => paramRe.test(segment)

const isWildcard = (segment: string) => segment == '*'

const rankRoute = (route: Route, index: number) => {
  const score = route.default
    ? 0
    : segmentize(route.path).reduce((score: number, segment: string) => {
        score += SEGMENT_POINTS

        if (isRootSegment(segment)) score += ROOT_POINTS
        else if (isDynamic(segment)) score += DYNAMIC_POINTS
        else if (isWildcard(segment)) score -= SEGMENT_POINTS + WILDCARD_PENALTY
        else score += STATIC_POINTS

        return score
      }, 0)

  return { route, score, index }
}

const rankRoutes = (routes: Route[]) =>
  routes
    .map(rankRoute)
    .sort((a, b) => (a.score < b.score ? 1 : a.score > b.score ? -1 : a.index - b.index))

const segmentize = (uri: string) =>
  uri
    // strip starting/ending slashes
    .replace(/(^\/+|\/+$)/g, '')
    .split('/')

const addQuery = (pathname: string, query: string) => pathname + (query ? `?${query}` : '')

const reservedNames = ['uri', 'path']

export {
  insertParams,
  invariant,
  isProduction,
  match,
  pick,
  resolve,
  shouldNavigate,
  startsWith,
  stripSlashes,
  validateRedirect
}
