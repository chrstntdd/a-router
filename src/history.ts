import { HistorySource, HistoryListener, EventHandler } from './types'

/**
 * @description To wrap a history source
 */
const getLocation = (source: Window | HistorySource) => ({
  ...source.location,
  state: source.history.state,
  key: (source.history.state && source.history.state.key) || 'initial'
})

const createHistory = (source: Window | HistorySource) => {
  let listeners: any[] = []
  let location = getLocation(source)
  let transitioning = false
  let resolveTransition = () => {}

  return {
    get location() {
      return location
    },

    get transitioning() {
      return transitioning
    },

    onTransitionComplete() {
      transitioning = false
      resolveTransition()
    },

    listen(listener: HistoryListener) {
      listeners.push(listener)

      const popstateListener = () => {
        location = getLocation(source)
        listener({ location, action: 'POP' })
      }

      source.addEventListener('popstate', popstateListener)

      return () => {
        source.removeEventListener('popstate', popstateListener)
        listeners = listeners.filter(fn => fn !== listener)
      }
    },

    navigate(to: string, { state = {}, replace = false } = {}) {
      state = { ...state, key: Date.now().toString() }

      // try...catch iOS Safari limits to 100 pushState calls
      try {
        if (transitioning || replace) {
          // @ts-ignore not updating the title of the page
          source.history.replaceState(state, null, to)
        } else {
          // @ts-ignore not updating the title of the page
          source.history.pushState(state, null, to)
        }
      } catch (e) {
        source.location[replace ? 'replace' : 'assign'](to)
      }

      location = getLocation(source)
      transitioning = true
      const transition = new Promise(res => (resolveTransition = res))
      listeners.forEach(listener => listener({ location, action: 'PUSH' }))
      return transition
    }
  }
}

/**
 * @description Stores history entries in memory for testing or other platforms like Native
 */
const createMemorySource = (initialPathname = '/'): HistorySource => {
  const stack = [{ pathname: initialPathname, search: '' }]
  const states: any[] = []
  let index = 0

  return {
    // @ts-ignore memory location is not fully spec compliant ¯\_(ツ)_/¯
    get location() {
      return stack[index]
    },
    addEventListener(name: string, fn: EventHandler) {},
    removeEventListener(name: string, fn: EventHandler) {},
    history: {
      get entries() {
        return stack
      },
      get index() {
        return index
      },
      get state() {
        return states[index]
      },
      pushState(state: any, _: never, uri: string) {
        const [pathname, search = ''] = uri.split('?')
        index++
        stack.push({ pathname, search })
        states.push(state)
      },
      replaceState(state: any, _: never, uri: string) {
        const [pathname, search = ''] = uri.split('?')
        stack[index] = { pathname, search }
        states[index] = state
      }
    }
  }
}

/**
 * @description To retrieve a history source. Uses `window.history`
 * if available, but falls back to using a memory history that
 * mirrors the same API
 */
const getSource = () => (typeof window === 'undefined' ? createMemorySource() : window)

const globalHistory = createHistory(getSource())
const { navigate } = globalHistory

export { globalHistory, navigate, createHistory, createMemorySource }
