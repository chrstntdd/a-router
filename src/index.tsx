import React from 'react'
import invariant from 'tiny-invariant'
import { unstable_scheduleCallback as defer } from 'scheduler'

/* CLONE/FORK OF https://github.com/reach/router */

import {
  insertParams,
  match,
  pick,
  resolve,
  Route,
  shouldNavigate,
  startsWith,
  stripSlashes,
  validateRedirect
} from './helpers'
import { createHistory, createMemorySource, globalHistory, navigate } from './history'

const createNamedContext = (name: string, defaultValue?: any): React.Context<any> => {
  const Ctx = React.createContext(defaultValue)
  //@ts-ignore
  Ctx.Consumer.displayName = `${name}.Consumer`
  //@ts-ignore
  Ctx.Provider.displayName = `${name}.Provider`

  return Ctx
}

// Sets baseuri and basepath for nested routers and links
let BaseContext = createNamedContext('Base', { baseuri: '/', basepath: '/' })

/**
 * @description Main Router component that connects the matched Component to
 * the contexts.
 */
let Router = props => {
  let baseContext = React.useContext(BaseContext)

  return (
    <Location>
      {locationContext => <RouterImpl {...baseContext} {...locationContext} {...props} />}
    </Location>
  )
}

interface PRouterImpl {
  basepath: any
  baseuri: any
  component: any
  location: any
  navigate: any
  children: any
  primary?: boolean
}

let RouterImpl = (props: PRouterImpl) => {
  let {
    basepath,
    baseuri,
    children,
    component = 'div',
    location,
    navigate,
    primary = true,
    ...domProps
  } = props

  let routes = React.Children.map(children, createRoute(basepath))

  let match = pick(routes, location.pathname)

  if (match) {
    let {
      params,
      uri,
      route,
      route: { value: element }
    } = match

    // remove the /* from the end for child routes relative paths
    basepath = route.default ? basepath : route.path.replace(/\*$/, '')

    let props = {
      ...params,
      uri,
      location,
      navigate: (to, options) => navigate(resolve(to, uri), options)
    }

    let clone = React.cloneElement(
      element,
      props,
      element.props.children ? (
        <Router primary={primary}>{element.props.children}</Router>
      ) : (
        undefined
      )
    )

    /* using 'div' for < 16.3 support */
    let FocusWrapper = primary ? FocusHandler : component
    /* don't pass any props to 'div' */
    let wrapperProps = primary ? { uri, location, component, ...domProps } : domProps

    return (
      <BaseContext.Provider value={{ baseuri: uri, basepath }}>
        <FocusWrapper {...wrapperProps}>{clone}</FocusWrapper>
      </BaseContext.Provider>
    )
  } else {
    return null
  }
}

let FocusContext = createNamedContext('Focus')

interface FocusHandlerProps {
  children: any
  component: any
  location: any
  uri: string
}

let FocusHandler = ({ uri, location, component, ...domProps }: FocusHandlerProps) => {
  let requestFocus = React.useContext(FocusContext)

  return (
    <FocusHandlerImpl
      {...domProps}
      component={component}
      requestFocus={requestFocus}
      uri={uri}
      location={location}
    />
  )
}

interface PFocusHandlerImpl {
  component: any
  children: any
  requestFocus: (any) => void
  uri: any
  location: any
  role?: string
  style?: object
}
interface SFocusHandlerImpl {
  shouldFocus?: boolean | null
}
// don't focus on initial render
let initialRender = true
let focusHandlerCount = 0

function FocusHandlerImpl(props: PFocusHandlerImpl) {
  const [state, setState] = React.useState({ shouldFocus: null })
  const compEl = React.useRef(null)

  // cDM && cWU
  React.useEffect(() => {
    focus()
    focusHandlerCount++
    setState({ shouldFocus: true, ...props })

    // cleanup
    return () => {
      focusHandlerCount--
      if (focusHandlerCount === 0) {
        initialRender = true
      }
    }
  }, [])

  // cDU
  React.useEffect(
    () => {
      if (state.shouldFocus) {
        focus()
      }
    },
    [props.location]
  )

  function focus() {
    if (process.env.NODE_ENV === 'test') {
      // getting cannot read property focus of null in the tests
      // and that bit of global `initialRender` state causes problems
      // should probably figure it out!
      return
    }

    let { requestFocus } = props

    requestFocus
      ? requestFocus(compEl.current)
      : initialRender
      ? (initialRender = false)
      : !compEl.current.contains(document.activeElement) && compEl.current.focus()
  }

  let internalRequestFocus = node => {
    !state.shouldFocus && node.focus()
  }

  let {
    children,
    component: Comp = 'div',
    location,
    requestFocus,
    role = 'group',
    style,
    uri,
    ...domProps
  } = props

  return (
    <Comp
      ref={compEl}
      role={role}
      style={{ outline: 'none', ...style }}
      tabIndex="-1"
      {...domProps}
    >
      <FocusContext.Provider value={internalRequestFocus}>{props.children}</FocusContext.Provider>
    </Comp>
  )
}

let Match = ({ path, children }) => {
  let { baseuri } = React.useContext(BaseContext)

  return (
    <Location>
      {({ navigate, location }) => {
        let resolvedPath = resolve(path, baseuri)
        let result = match(resolvedPath, location.pathname)

        return children({
          navigate,
          location,
          match: result
            ? {
                ...result.params,
                uri: result.uri,
                path
              }
            : null
        })
      }}
    </Location>
  )
}

/**
 * Location START
 */

// Location Context/Provider
let LocationContext = createNamedContext('Location')

// sets up a listener if there isn't one already so apps don't need to be
// wrapped in some top level provider
let Location = ({ children }) => {
  let locationContext = React.useContext(LocationContext)

  return locationContext ? (
    children(locationContext)
  ) : (
    <LocationProvider>{children}</LocationProvider>
  )
}

interface PLocationProvider {
  history?: any
  children?: (any) => React.ReactNode
}
interface SLocationProvider {
  context: {
    navigate: any
    location: any
  }
  refs: {
    unlisten: any
  }
}

LocationProvider.defaultProps = {
  history: globalHistory
}

function LocationProvider(props: PLocationProvider) {
  const [locationState, setLocationState] = React.useState({
    context: getContext(),
    refs: { unlisten: null }
  })

  let unmounted: boolean | null = null

  React.useEffect(() => {
    locationState.refs.unlisten = props.history.listen(() => {
      Promise.resolve().then(() => {
        defer(() => {
          if (!unmounted) {
            //@ts-ignore
            setLocationState({ context: getContext() })
          }
        })
      })
    })

    return () => {
      unmounted = true
      locationState.refs.unlisten()
    }
  }, [])

  function getContext() {
    let {
      history: { navigate, location }
    } = props

    return { navigate, location }
  }

  try {
    return (
      <LocationContext.Provider value={locationState.context}>
        {typeof props.children === 'function'
          ? props.children(locationState.context)
          : props.children || null}
      </LocationContext.Provider>
    )
  } catch (error) {
    // componentDidCatch...?
    if (isRedirect(error)) {
      props.history.navigate(error.uri, { replace: true })
    } else {
      throw error
    }
  }
}

/**
 *
 * @description When you render a <Redirect/> a redirect request is thrown,
 * preventing react from rendering the whole tree when we donâ€™t want to do
 * that work anyway.
 *
 * To enable SSR, wrap the top level <App /> component with this component
 * and pass it the url that exists on the request object of whichever node
 * framework is being used.
 */
let ServerLocation = ({ url, children }) => (
  <LocationContext.Provider
    value={{
      location: { pathname: url },
      navigate: () => {
        throw new Error("You can't call navigate on the server.")
      }
    }}
  >
    {children}
  </LocationContext.Provider>
)

/**
 * Location END ////////////////////////////////////////////////
 */

/**
 * Redirect START ///////////////////////////////////////////////
 */

interface IRedirectRequest {
  uri: string
}

function RedirectRequest(this: IRedirectRequest, uri: string) {
  this.uri = uri
}

let isRedirect = o => o instanceof RedirectRequest

let redirectTo = to => {
  throw new RedirectRequest(to)
}

interface PRedirectImpl {
  from: string
  navigate: (toPath: string, state: { replace: boolean; state: object }) => any
  noThrow: boolean
  replace: boolean
  state: any
  to: string
}

let RedirectImpl = (props: PRedirectImpl) => {
  let { navigate, to, from, replace, state, noThrow, ...restProps } = props
  if (!noThrow) redirectTo(insertParams(to, restProps))

  return null
}

let Redirect = props => {
  let locationContext = React.useContext(LocationContext)

  return <RedirectImpl {...locationContext} {...props} />
}

interface LinkPropGetter {
  isCurrent: boolean
  isPartiallyCurrent: boolean
  href: string
  location: any
}

interface LinkProps {
  to: string
  innerRef?: any
  state?: any
  replace?: () => any
  getProps?: (x: LinkPropGetter) => any
}

let Link: React.ComponentType<LinkProps & React.HTMLProps<HTMLAnchorElement>> = props => {
  let { baseuri } = React.useContext(BaseContext)
  let linkRef = React.useRef(null)

  return (
    <Location>
      {({ location, navigate }) => {
        let { to, state, replace, getProps = () => {}, ...anchorProps } = props
        let href = resolve(to, baseuri)
        let isCurrent = location.pathname === href
        let isPartiallyCurrent = startsWith(location.pathname, href)

        return (
          <a
            ref={linkRef}
            aria-current={isCurrent ? 'page' : undefined}
            {...anchorProps}
            {...getProps({ isCurrent, isPartiallyCurrent, href, location })}
            href={href}
            onClick={event => {
              if (anchorProps.onClick) anchorProps.onClick(event)
              if (shouldNavigate(event)) {
                event.preventDefault()
                navigate(href, { state, replace })
              }
            }}
          />
        )
      }}
    </Location>
  )
}

/**
 * Link END //////////////////////////////////////////////////////////////////
 */

/**
 * Extras
 */

let createRoute = basepath => (element): Route => {
  if (!element) return null

  if (process.env.NODE_ENV !== 'production') {
    invariant(
      element.props.path || element.props.default || element.type === Redirect,
      `<Router>: Children of <Router> must have a \`path\` or \`default\` prop, or be a \`<Redirect>\`. None found on element type \`${
        element.type
      }\``
    )

    invariant(
      !(element.type === Redirect && (!element.props.from || !element.props.to)),
      `<Redirect from="${element.props.from} to="${
        element.props.to
      }"/> requires both "from" and "to" props when inside a <Router>.`
    )

    invariant(
      !(element.type === Redirect && !validateRedirect(element.props.from, element.props.to)),
      `<Redirect from="${element.props.from} to="${
        element.props.to
      }"/> has mismatched dynamic segments, ensure both paths have the exact same dynamic segments.`
    )
  }

  if (element.props.default) {
    return { value: element, default: true }
  }

  let elementPath = element.type === Redirect ? element.props.from : element.props.path

  let path =
    elementPath === '/' ? basepath : `${stripSlashes(basepath)}/${stripSlashes(elementPath)}`

  return {
    value: element,
    default: element.props.default,
    path: element.props.children ? `${stripSlashes(path)}/*` : path
  }
}

export {
  BaseContext,
  createHistory,
  createMemorySource,
  globalHistory,
  Link,
  Location,
  LocationProvider,
  Match,
  navigate,
  Redirect,
  Router,
  ServerLocation
}
