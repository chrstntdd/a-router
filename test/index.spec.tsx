import * as React from 'react'
import 'jest-dom/extend-expect'
import { render, cleanup, fireEvent, wait } from 'react-testing-library'

import { createHistory, createMemorySource, LocationProvider, Router, Link } from '../src'

jest.useFakeTimers()

const runWithNavigation = ({ element, pathname = '/' }) => {
  const history = createHistory(createMemorySource(pathname))
  const wrapper = render(<LocationProvider history={history}>{element}</LocationProvider>)
  const snapshot = () => {
    expect(wrapper.container.firstChild).toMatchSnapshot()
  }

  return { history, snapshot, wrapper }
}

const snapshot = ({ pathname, element }) => {
  const testHistory = createHistory(createMemorySource(pathname))
  const wrapper = render(<LocationProvider history={testHistory}>{element}</LocationProvider>)

  expect(wrapper.container.firstChild).toMatchSnapshot()
}

const NotFound = () => <div>404 page</div>
const Home = () => (
  <div>
    Home
    <br />
    <Link to="/">Current page</Link>
    <Link data-testid="dash-anchor" to="/dash">
      dash
    </Link>
  </div>
)
const Dash = ({ children }) => (
  <div>
    Dash
    <br />
    {children}
    <br />
    <Link to="/dash">Current page</Link>
    <Link data-testid="home-anchor" to="/">
      Home
    </Link>
  </div>
)
const Group = ({ groupId, children }) => (
  <div>
    Group: {groupId}
    {children}
  </div>
)
const PropsPrinter = props => <pre>{JSON.stringify(props, null, 2)}</pre>
const Reports = ({ children }) => <div>Reports {children}</div>
const AnnualReport = () => <div>Annual Report</div>

describe('sus-router', () => {
  afterEach(cleanup)

  describe('smoke', () => {
    it(`renders the root component at "/"`, () => {
      snapshot({
        pathname: '/',
        element: (
          <Router>
            <Home path="/" />
            <Dash path="/dash" />
          </Router>
        )
      })
    })

    it('renders at a path', () => {
      snapshot({
        pathname: '/dash',
        element: (
          <Router>
            <Home path="/" />
            <Dash path="/dash" />
          </Router>
        )
      })
    })

    it('renders a fallback route', () => {
      const {
        wrapper: { getByText }
      } = runWithNavigation({
        pathname: '/notfoundroute',
        element: (
          <Router>
            <Home path="/" />
            <Dash path="/dash" />
            <NotFound default />
          </Router>
        )
      })

      expect(getByText(/404 page/i)).toBeInTheDocument()
    })

    it('renders a lazy component', async () => {
      const TestAsyncComponent = React.lazy(() => import('./FakeLazy'))

      const {
        wrapper: { getByText }
      } = runWithNavigation({
        pathname: '/lazy',
        element: (
          <React.Suspense fallback={<div />}>
            <Router>
              <Home path="/" />
              <Dash path="/dash" />
              <TestAsyncComponent path="/lazy" />
              <NotFound default />
            </Router>
          </React.Suspense>
        )
      })

      await wait()
      expect(getByText(/lazy/i)).toBeInTheDocument()
    })
  })

  describe('using links', () => {
    it('should have functioning links, lol', async () => {
      const {
        history,
        wrapper: { getByTestId, getByText, debug }
      } = runWithNavigation({
        pathname: '/',
        element: (
          <Router>
            <Home path="/" />
            <Dash path="/dash" />
          </Router>
        )
      })

      expect(getByText(/home/i)).toBeInTheDocument()
      expect(history.location.pathname).toEqual('/')

      fireEvent.click(getByTestId('dash-anchor'))

      // TODO find way of asserting that the page has changed without inspecting the history
      // const dashboardPage = await waitForElement(() => getByTestId('home-anchor'))

      expect(history.location.pathname).toEqual('/dash')
    })

    it('should **NOT** update anything when clicking on a link to the current url', () => {
      const {
        history,
        wrapper: { getByText }
      } = runWithNavigation({
        pathname: '/',
        element: (
          <Router>
            <Home path="/" />
            <Dash path="/dash" />
          </Router>
        )
      })

      const initialHistory = history.location

      expect(history.location.pathname).toEqual('/')
      expect(getByText(/home/i)).toBeInTheDocument()

      fireEvent.click(getByText(/current page/i))

      expect(initialHistory).toEqual(history.location)
      expect(getByText(/home/i)).toBeInTheDocument()
      expect(history.location.pathname).toEqual('/')
    })
  })
})
