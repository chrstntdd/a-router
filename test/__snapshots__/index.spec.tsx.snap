// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sus-router smoke renders at a path 1`] = `
<div
  role="group"
  tabindex="-1"
>
  <div>
    Dash
    <br />
    <br />
    <a
      aria-current="page"
      href="/dash"
    >
      Current page
    </a>
    <a
      data-testid="home-anchor"
      href="/"
    >
      Home
    </a>
  </div>
</div>
`;

exports[`sus-router smoke renders the root component at "/" 1`] = `
<div
  role="group"
  tabindex="-1"
>
  <div>
    Home
    <br />
    <a
      aria-current="page"
      href="/"
    >
      Current page
    </a>
    <a
      data-testid="dash-anchor"
      href="/dash"
    >
      dash
    </a>
  </div>
</div>
`;
